File: static/src/js/edi_wall.js
/** @odoo-module **/
// mml_edi/static/src/js/edi_wall.js
//
// EDI operations WALL DISPLAY — a 1920x1080 dark ops radiator for a TV browser
// session. Recreates design_handoff "Wall Display.dc.html" as a production OWL
// 2 client action (tag `edi_wall`), continuing the 3PL / ROQ visual language.
// Read-only: no interaction, no filter — it just refreshes on the module's
// 5-minute cadence (README EDI screen 7). Layout:
//   - header: app mark + "MML EDI · order exchange · operations", live dot, clock
//   - 4 alarm tiles (blocking reviews / failed ORDRSP / unknown stores / poll
//     freshness) that GLOW when non-zero (calm green -> amber -> red by count)
//   - big inbound pipeline ramp (Received -> Parsed -> Orders -> In review ->
//     Acknowledged) with today's header line
//   - service-level bars (Auto-approval / ORDRSP on-time / Clean-parse) with a
//     target marker, + a partner strip
//   - live exchange feed ticker
//
// One batched RPC — edi.wall.get_wall_summary() — returns the whole board. The
// FIXED dark radiator palette is applied in BOTH Odoo colour schemes (the wall
// reads the same on a TV regardless of the operator's theme), so this screen
// uses its own hardcoded palette rather than the --edi-* token bridge.

import { registry } from "@web/core/registry";
import { Component, useState, onWillStart, onMounted, onWillUnmount } from "@odoo/owl";
import { useService } from "@web/core/utils/hooks";

const REFRESH_MS = 5 * 60 * 1000; // keep the module's 5-min live-refresh cadence
const CLOCK_MS = 15 * 1000;       // wall clock tick

// Alarm tone -> tile colour / inner glow / border (fixed, both schemes).
const ALARM_TONES = {
    calm: { color: "#46A758", glow: "transparent", border: "#212B42" },
    amber: { color: "#F5A524", glow: "rgba(245,165,36,.14)", border: "#5A4620" },
    red: { color: "#E5484D", glow: "rgba(229,72,77,.16)", border: "#5A2327" },
};

// Inbound-pipeline ramp: the wall's dark-radiator stage palette (README
// "Pipeline / stage ramp" + the prototype's dark boxes). Position-ordered.
const STAGE_STYLES = [
    { bg: "#12305A", fg: "#BCD6F5" },   // Received
    { bg: "#1C4A86", fg: "#DCEAFB" },   // Parsed
    { bg: "#2E7CD6", fg: "#fff" },      // Orders created
    { bg: "#4A3512", fg: "#F5C97A" },   // In review (dwell amber)
    { bg: "#173A22", fg: "#7FD79A" },   // Acknowledged
];

// Service-level bar tone -> stroke colour.
const BAR_COLORS = { green: "#46A758", amber: "#F5A524", red: "#E5484D", none: "#5C6784" };

function nowClock() {
    const d = new Date();
    const hh = String(d.getHours()).padStart(2, "0");
    const mm = String(d.getMinutes()).padStart(2, "0");
    return hh + ":" + mm;
}

class EdiWall extends Component {
    static template = "mml_edi.EdiWall";
    static props = ["*"];

    setup() {
        this.orm = useService("orm");
        this.state = useState({
            loading: true,
            error: null,
            summary: null,
            clock: nowClock(),
        });
        this._refreshInterval = null;
        this._clockInterval = null;

        onWillStart(async () => {
            await this._load();
        });
        onMounted(() => {
            this._refreshInterval = setInterval(() => this._load(), REFRESH_MS);
            this._clockInterval = setInterval(() => { this.state.clock = nowClock(); }, CLOCK_MS);
        });
        onWillUnmount(() => {
            if (this._refreshInterval) clearInterval(this._refreshInterval);
            if (this._clockInterval) clearInterval(this._clockInterval);
        });
    }

    async _load() {
        try {
            const summary = await this.orm.call("edi.wall", "get_wall_summary", []);
            Object.assign(this.state, { summary, loading: false, error: null });
        } catch (e) {
            Object.assign(this.state, { error: "Couldn't load the wall", loading: false });
        }
    }

    get dataAvailable() {
        return !!(this.state.summary && this.state.summary.data_available);
    }

    // ---- alarm tiles ---------------------------------------------------------

    alarmTiles() {
        const alarms = (this.state.summary && this.state.summary.alarms) || [];
        return alarms.map((a) => {
            const tone = ALARM_TONES[a.tone] || ALARM_TONES.calm;
            return {
                key: a.key,
                label: a.label,
                value: a.value,
                sub: a.sub,
                valueStyle: "color:" + tone.color,
                tileStyle:
                    "border:1px solid " + tone.border +
                    ";box-shadow:inset 0 0 40px " + tone.glow,
            };
        });
    }

    // ---- inbound pipeline ----------------------------------------------------

    get todayLine() {
        const t = (this.state.summary && this.state.summary.today) || {};
        return [
            (t.files || 0) + " files",
            (t.pos || 0) + " POs",
            (t.store_orders || 0) + " store-orders",
            (t.acks || 0) + " ORDRSP sent",
        ].join(" · ");
    }

    pipelineStages() {
        const p = (this.state.summary && this.state.summary.pipeline) || {};
        const notes = p.notes || {};
        const defs = [
            { key: "received", label: "Received", value: p.received },
            { key: "parsed", label: "Parsed", value: p.parsed },
            { key: "orders_created", label: "Orders created", value: p.orders_created },
            { key: "in_review", label: "In review", value: p.in_review },
            { key: "acknowledged", label: "Acknowledged", value: p.acknowledged },
        ];
        return defs.map((d, idx) => {
            const st = STAGE_STYLES[idx];
            return {
                key: d.key,
                label: d.label,
                value: typeof d.value === "number" ? d.value : 0,
                note: notes[d.key] || "",
                style: "background:" + st.bg + ";color:" + st.fg,
                arrow: idx < defs.length - 1,
            };
        });
    }

    // ---- service-level bars --------------------------------------------------

    serviceBars() {
        const bars = (this.state.summary && this.state.summary.service_levels) || [];
        return bars.map((b, idx) => {
            const color = BAR_COLORS[b.tone] || BAR_COLORS.none;
            return {
                key: "bar" + idx,
                label: b.label,
                value: b.value,
                target: b.target,
                valueStyle: "color:" + color,
                fillStyle: "width:" + b.pct + ";background:" + color,
                markStyle: "left:" + b.mark,
            };
        });
    }

    // ---- partner strip -------------------------------------------------------

    partnerStrip() {
        const partners = (this.state.summary && this.state.summary.partners) || [];
        return partners.map((p) => ({
            id: p.id,
            name: p.name,
            sub: p.sub,
            dotStyle: "background:" + p.dot,
        }));
    }

    // ---- live feed -----------------------------------------------------------

    feedRows() {
        const feed = (this.state.summary && this.state.summary.feed) || [];
        return feed.map((ev) => ({
            key: "f" + ev.id,
            time: ev.time,
            tag: ev.tag,
            tagStyle: "background:" + ev.color,
            text: ev.text,
        }));
    }

    // ---- poll dot ------------------------------------------------------------

    get pollFresh() {
        return !!(this.state.summary && this.state.summary.poll && this.state.summary.poll.last_push_fresh);
    }
}

registry.category("actions").add("edi_wall", EdiWall);

export { EdiWall };

File: models/edi_wall.py
# mml_edi/models/edi_wall.py
"""EDI operations wall-display computation service.

A 1920x1080 dark ops radiator for a TV browser session (README EDI screen 7,
``Wall Display.dc.html``). The prototype is recreated as a read-only OWL client
action (``edi_wall``) that auto-refreshes on the module's 5-minute cadence — no
interaction.

AbstractModel — no DB table. The whole board is one batched ``@api.model`` RPC,
``edi.wall.get_wall_summary(partner_code)``, composed largely from the
``edi.dashboard`` helpers (pipeline, today totals, KPI window, live feed,
partner health, attention triage) so the wall and the landing dashboard never
drift. The wall adds only its own shaping: the four alarm tiles (calm/amber/red
by count), the three service-level bars with a target marker, the partner
strip, and the poll-freshness reading.

Honesty: the wall reports only what the EDI models record. The alarm counts and
service levels come straight from ``edi.order.review`` / ``edi.log`` — nothing
is fabricated. The FIXED dark radiator palette (README shared-theme "Wall
Display" row) is applied in the frontend in BOTH Odoo colour schemes, so the
board reads the same on a TV regardless of the operator's theme.
"""
import logging
from datetime import timedelta

from odoo import api, fields, models

from ..services.wall_format import alarm_count_tone, poll_tone, service_bar
from .edi_dashboard import _HEALTH_DOTS, _pct

_logger = logging.getLogger(__name__)


class EdiWall(models.AbstractModel):
    """Wall-display computation service for the EDI OWL radiator.

    AbstractModel (no table). Every method is ``@api.model``, called from the
    frontend via ``this.orm.call('edi.wall', 'get_wall_summary', [], {...})``.
    """

    _name = "edi.wall"
    _description = "EDI Operations Wall Display"
    _table = "edi_wall"

    # ---- main entry point ----------------------------------------------------

    @api.model
    def get_wall_summary(self, partner_code=None):
        """Return the full wall-display payload, optionally scoped to a partner.

        Payload keys:
          - ``data_available`` — False before any inbound activity (the wall
            then shows a "waiting for the first poll" state, not a false all-green
            radiator).
          - ``alarms`` — the four alarm tiles (blocking reviews, failed ORDRSP,
            unknown stores, poll freshness) with a ``tone`` driving the glow.
          - ``pipeline`` / ``today`` — the inbound ramp + its header line.
          - ``service_levels`` — the three service-level bars (auto-approval,
            ORDRSP on-time, clean-parse) with value / target / marker / tone.
          - ``partners`` — the partner strip (dot + format + in-review count).
          - ``feed`` — the recent typed exchange events.
          - ``poll`` — the poll-freshness dot (``last_push`` / ``last_push_fresh``).
        """
        dash = self.env["edi.dashboard"]
        now = fields.Datetime.now()
        today_start = dash._local_today_start(now)
        window_start = now - timedelta(days=30)
        targets = dash.get_kpi_targets()

        # One batched attention pass feeds every alarm count (no per-alarm search).
        attention = dash._attention_items(now, window_start, partner_code)
        kpis = dash._kpis(now, window_start, targets, partner_code)

        return {
            "data_available": dash._data_available(partner_code),
            "alarms": self._alarms(now, attention, partner_code),
            "pipeline": dash._pipeline(today_start, partner_code),
            "today": dash._today_totals(today_start, partner_code),
            "service_levels": self._service_levels(kpis, targets, window_start, partner_code),
            "partners": self._partner_strip(partner_code),
            "feed": dash._live_feed(now, partner_code, limit=10),
            "poll": dash._feed_freshness(now, partner_code),
        }

    # ---- alarm tiles ---------------------------------------------------------

    @api.model
    def _alarms(self, now, attention, partner_code=None):
        """The four wall alarm tiles, derived from the batched attention pass +
        one poll-freshness read.

        ``attention`` (from ``edi.dashboard._attention_items``) already groups the
        actionable triage rows, so the blocking / failed-ORDRSP / unknown-store
        counts come from it with no extra query.
        """
        blocking = sum(1 for i in attention if i["kind"] == "blocking")
        failed_ack = sum(1 for i in attention if i["kind"] == "ack_failed")
        unknown = sum(
            1 for i in attention
            if i["kind"] == "warning" and "map_store" in i.get("actions", [])
        )

        fresh_min = self._poll_freshness_minutes(now, partner_code)
        poll_value = "—" if fresh_min is None else str(fresh_min)
        poll_sub = "no poll yet" if fresh_min is None else "min since last poll"

        return [
            {"key": "blocking", "label": "Blocking reviews", "value": str(blocking),
             "sub": "need a decision", "tone": alarm_count_tone(blocking)},
            {"key": "failed_ack", "label": "Failed ORDRSP", "value": str(failed_ack),
             "sub": "upload failed · retry armed", "tone": alarm_count_tone(failed_ack)},
            {"key": "unknown", "label": "Unknown stores", "value": str(unknown),
             "sub": "unmapped store code", "tone": alarm_count_tone(unknown)},
            {"key": "poll", "label": "Poll freshness", "value": poll_value,
             "sub": poll_sub, "tone": poll_tone(fresh_min)},
        ]

    @api.model
    def _poll_freshness_minutes(self, now, partner_code=None):
        """Whole minutes since the newest inbound poll, or None if never polled."""
        dash = self.env["edi.dashboard"]
        last = self.env["edi.log"].search(
            dash._partner_domain(partner_code) + [("direction", "=", "inbound")],
            order="timestamp desc", limit=1)
        if not last or not last.timestamp:
            return None
        return int((now - last.timestamp).total_seconds() // 60)

    # ---- service-level bars --------------------------------------------------

    @api.model
    def _clean_parse_targets(self):
        """Clean-parse-rate target + amber floor (config-overridable defaults)."""
        ICP = self.env["ir.config_parameter"].sudo()

        def _f(key, default):
            return float(ICP.get_param("mml_edi." + key) or default)

        return (_f("kpi_clean_parse_target", "98"), _f("kpi_clean_parse_amber", "95"))

    @api.model
    def _clean_parse_rate(self, window_start, partner_code=None):
        """Successful file parses / all parse attempts over the 30d window.

        A parse error (a malformed inbound file that could not be read) drags
        this off 100% — an honest quality signal, not a fabricated one. None when
        nothing has parsed.
        """
        Log = self.env["edi.log"]
        pdom = self.env["edi.dashboard"]._partner_domain(partner_code)
        ok = Log.search_count(
            pdom + [("event_type", "=", "file_parse"), ("status", "=", "success"),
                    ("timestamp", ">=", window_start)])
        err = Log.search_count(
            pdom + [("event_type", "=", "file_parse"), ("status", "=", "error"),
                    ("timestamp", ">=", window_start)])
        return _pct(ok, ok + err)

    @api.model
    def _service_levels(self, kpis, targets, window_start, partner_code=None):
        """The three wall service-level bars (auto-approval, ORDRSP on-time,
        clean-parse). Auto-approval and ORDRSP on-time reuse the dashboard KPI
        window; clean-parse is computed here."""
        clean_target, clean_amber = self._clean_parse_targets()
        clean = self._clean_parse_rate(window_start, partner_code)
        return [
            service_bar("Auto-approval rate", kpis["auto_approval"]["value"],
                        targets["auto_approval_target"], targets["auto_approval_amber"]),
            service_bar("ORDRSP on-time", kpis["ordrsp_on_time"]["value"],
                        targets["ordrsp_on_time_target"], targets["ordrsp_on_time_amber"]),
            service_bar("Clean-parse rate", clean, clean_target, clean_amber),
        ]

    # ---- partner strip -------------------------------------------------------

    @api.model
    def _partner_strip(self, partner_code=None):
        """Partner strip: name + health dot + 'format · N in review' per partner."""
        partners = self.env["edi.dashboard"]._partner_for_code(partner_code).sorted("name")
        rows = []
        for p in partners:
            health = p.health_state or "never"
            fmt = dict(p._fields["edi_format"].selection).get(p.edi_format, p.edi_format or "")
            rows.append({
                "id": p.id,
                "name": p.name,
                "dot": _HEALTH_DOTS.get(health, "#adb5bd"),
                "sub": "%s · %d in review" % (fmt, p.pending_review_count),
            })
        return rows
